source/extensions/filters/http/oauth2/filter.cc

```diff
@@ -173,19 +179,22 @@ std::string encodeHmac(const std::vector<uint8_t>& secret, absl::string_view dom
   return encodeHmacBase64(secret, domain, expires, token, id_token, refresh_token);
 }
 
-// Generates a nonce based on the current time
-std::string generateFixedLengthNonce(TimeSource& time_source) {
-  constexpr size_t length = 16;
-
-  std::string nonce = fmt::format("{}", time_source.systemTime().time_since_epoch().count());
-
-  if (nonce.length() < length) {
-    nonce.append(length - nonce.length(), '0');
-  } else if (nonce.length() > length) {
-    nonce = nonce.substr(0, length);
-  }
+// Generates a non-guessable nonce that can be used to prevent CSRF attacks.
+// The nonce is a base64 encoded SHA256 HMAC generated from the current timestamp and a secret.
+std::string generateNonce(absl::string_view hmac_secret, TimeSource& time_source) {
+  std::vector<uint8_t> hmac_secret_vec(hmac_secret.begin(), hmac_secret.end());
+  std::string timestamp = fmt::format("{}", time_source.systemTime().time_since_epoch().count());
+  return generateHmacBase64(hmac_secret_vec, timestamp);
+}
```

Comment:
I will prefer to keep the original nonce generating logic if there is no strong reason.

Reply:
Change the nonce from a plain timestamp to a SHA256 HMAC, generated using the hmac secret and the timestamp, to enhance its security robustness. This addresses the "1. Generation" in this issue https://github.com/envoyproxy/envoy/issues/36276

Reply:
After discussed with @wbpcode , changed the nonce to a random, which is light weight compared with HMAC and also secure enoug.